### DISCUSSION TITLE
Change button configurations

### DISCUSSION THREAD

**Discussion by jerzyks** [outdoorbits/little-backup-box#231] [Q&A]
Hi great job - CONGRATULATIONS, I bought gpio oled + buttons hat from waveshare and the display works perfectly, but I am not able to make the buttons work, in the script displaymenu.py I changed the definitions for the menu 2 buttons to the ones on the board but none of the buttons work and I don't know why, can someone see what is the reason?

[https://www.waveshare.com/1.3inch-oled-hat.htm](url)

Raspberry Pi PIN (BCM) buttons HAT
`Joystick Up 	P6 	Joystick Up
Joystick Down 	P19 	Joystick Down
Joystick Left 	P5 	Joystick Left
Joystick Right 	P26 	Joystick Right
`

config displaymenu.py 
`  
                      '2':    {
                                        'up':           6,
                                        'down':         19,
                                        'left':         5,
                                        'right':        26
                                }
`

**Comment by outdoorbits** [comment]
Thank you. Two short ideas, I didn't have the time to check more exactly yet: 
1. Did you activate the buttons in the settings?
2.  Did you use GPIO naming our PIN numbers? Please use GPIO.

If this works, please tell me, I would integrate this as an option of button settings.

Using buttons you maybe better install the development version.

**Reply by jerzyks** [reply]
YES - I activated right second option 
NO - changing the numbering to GPIO did not change anything - any buttons do not react

The Settings HARDWARE menu does not confirm the change in GPIO numbering, or the description in the menu takes information from a different place than the displaymenu.py ??

![obraz](https://github.com/outdoorbits/little-backup-box/assets/18427637/70ff77d1-190e-49c7-afa0-1fd463ac7691)

**Comment by outdoorbits** [comment]
Could you please send the modifications you did at the code?

**Reply by jerzyks** [reply]
Here is modified file [http://trx.sldc.pl/LBB/displaymenu.py](url)

**Comment by outdoorbits** [comment]
Oh, I didn't see them image above on my mobile. Following the manual of your board, this should be the correct pinning:
Joystick Up 	        GPIO 6     = PIN 31
Joystick Down 	GPIO 19   = PIN 35
Joystick Left 	        GPIO 5 	  = PIN 29
Joystick Right 	GPIO 26   = PIN 37

Your modified code should be:
`'2':	{
					'up':		6,
					'down':		19,
					'left':		5,
					'right':	        26
}`

Am I wrong?

**Comment by jerzyks** [comment]
Well, buttons still doesnt works, and button assign looks strange watch on menu and script.

![obraz](https://github.com/outdoorbits/little-backup-box/assets/18427637/7238c3cd-1895-4950-9fe3-da5ecc07d57b)
Am I really editing a working script? maybe just installer??? Edited path below
`/home/pi/little-backup-box/scripts/displaymenu.py`

**Comment by outdoorbits** [comment]
1. Please edit /var/www/little-backup-box/displaymenu.py
2. Do not use the PIN numbers but the GPIO. Not 31 but 6, not 35 but 19 ...

**Comment by outdoorbits** [comment]
I added your configuration to the development branch. Just install it by
` curl -sSL https://raw.githubusercontent.com/outdoorbits/little-backup-box/development/install-little-backup-box.sh | bash -s -- development 2> install-error.log`

**Reply by jerzyks** [reply]
hey don't be mad at me but it doesn't work, i first tried to change the same settings in the file /var/www/little-backup-box/displaymenu.py as above and finally they started to be visible in the Configuration of the control buttons menu, but the buttons however, they didn't work.
I updated the software version issued by you, selected the driver version and changed the buttons to option no. 3. The Little backup box logo is permanently displayed on the display and the keys do not work. The display, however, reacts to changes in settings remotely, sorrrrrry :-(

**Comment by outdoorbits** [comment]
There is not a lot documentation on waveshare. But There is a github project: https://github.com/pangduckwai/PiDisplay

Here I found a difference. They setup GPIO `pull_up_down=GPIO.PUD_UP` while they are also waiting for GPIO.RISING. This doesn't make sense so me right now but have a try:

Please alter /var/www/little-backup-box/displaymenu.py in 
line 252 from
`GPIO.setup(pin, GPIO.IN, pull_up_down = GPIO.PUD_DOWN)`
to
`GPIO.setup(pin, GPIO.IN, pull_up_down = GPIO.PUD_UP)`

To activate changes you have to go to web ui settings and save to restart the display-driver. Or you just reboot.

**Comment by jerzyks** [comment]
YES my Master - it works perfectly !!!!!!!!!!! and it doesn't make any sense at all because in the waveshare documentation there is a ready to use python example that this is how it works - [https://www.waveshare.com/w/upload/5/53/1.3inch-OLED-HAT-Code.7z](url)

`GPIO.setup(KEY_UP_PIN,      GPIO.IN, pull_up_down=GPIO.PUD_UP)  `

Your project is especially close to me because I wrote a similar one, but not as professionally as you, because I'm not a programmer. My idea was simple - the first USB that was connected to the RPI was mass storage, and the next ones were external storage. The program on the USB drive also created UUID directories, but also checked whether the inserted card had such a number and if not, it first assigned it to the card from the UUID generator. And then, just by pressing a button, the copying took place. In addition, you could insert any number of cards for copying, they were all mounted to one directory and RSync did everything by itself. Unfortunately, I've been extremely busy in recent years and the project died :-(

Of course there was no web server or SQL etc...

Thank you very much for your help!

**Reply by outdoorbits** [reply]
;-) Thank you. I'll add this to settings, maybe next week.
If you'd like to bring your experience and contribute, warmly welcome!

**Reply by jerzyks** [reply]
Yes with pleasure! I need to take a closer look at how the whole project works. I'll be in touch soon! Thank You one again

**Comment by outdoorbits** [comment]
All the button settings from above are now integrated into settings. Could you please give feedback if it is working or not?

**Comment by jerzyks** [comment]
I edited working displaymenu.py and works perfect - thank You!

**Comment by outdoorbits** [comment]
It should work without editing the py files now.

**Comment by jerzyks** [comment]
so needs update to check it???

**Comment by outdoorbits** [comment]
Yes, please update and change hardware settings.

**Comment by jerzyks** [comment]
no, it doesnt work because:

- forgot `GPIO.setup(pin, GPIO.IN, pull_up_down = GPIO.PUD_UP)`
- non working buttons assigns  has `'2':	{ 'up':		6, 'down':		13, 'left':		19, 'right':	        26 }`


working buttons assigns is `'2':	{ 'up':		6, 'down':		19, 'left':		5, 'right':	        26 }`

**Reply by outdoorbits** [reply]
No, don't change the code. It's completely configurable in settings/hardware now.

**Comment by jerzyks** [comment]
I did NOT change code - just curl new sh and thats all. After that checked whats problem

![image](https://github.com/outdoorbits/little-backup-box/assets/18427637/6a6dc752-50d0-4e84-9b76-9b6f1bc2fe3c)

**Comment by outdoorbits** [comment]
Did you try to change button properties in settings/hardware? You also 
could define functions to your spare buttons on the board.

**Comment by jerzyks** [comment]
yes I did try on variant 1 and 2 but no luck. 
I don't now how define spare buttons and even don't have idea what functions.

**Comment by outdoorbits** [comment]
Below the selection of the variants there is a link to a config file. Example you'll find in the file.

**Comment by outdoorbits** [comment]
Did you set the edge detection and the resistor?

**Comment by jerzyks** [comment]
I know what edge detection and the resistor but this is orginal OLED HAT, nothing soldering by me :-) 
I didn't find this example in the config file, can you show what you mean? line number ?

**Comment by outdoorbits** [comment]
It's inside the settings page:
![Screenshot_20230910_083315](https://github.com/outdoorbits/little-backup-box/assets/45213053/aeddfd9c-aba7-4361-b957-c98cdc7b822b)
Just follow the link "Edit custom configuration file", edit the file as described inside the file, save, reload the settings page ans select your entry.

**Comment by jerzyks** [comment]
Works perfect - OLED and buttons - here is working setup...
![obraz](https://github.com/outdoorbits/little-backup-box/assets/18427637/3f8cdd4b-1fae-4d16-af50-02142faf91a4)
...only there will be a multi usb backup and it will be a PERFECT tool not only for a photographer :-)))

**Reply by roshin8** [reply]
Can you send what your custom configuration file? I want to use the same hardware as well :)

**Reply by jerzyks** [reply]
I do not have modified main version, only developer version so my file is other, but You can use developer version with above settings, or modify main version with this description: [https://github.com/outdoorbits/little-backup-box/discussions/231#discussioncomment-6897073](url)

**Reply by outdoorbits** [reply]
Why don't you just use development version and setup like the screenshot?

**Reply by roshin8** [reply]
thank you!!! This works!!!

**Comment by outdoorbits** [comment]
...only there will be a multi usb backup and it will be a PERFECT tool not only for a photographer :-)))

Hopefully now it is a "PERFECT tool not only for a photographer". ;-) At least, the multi USB backup is integrated into development version. Please test an give feedback!

**Comment by jerzyks** [comment]
Ok, I don't think I know how USB multicopy works. First, I select the BACKUP option, then insert TARGET, then SOURCE(s), the display shows the number of files from both USBs, but only the contents of one USB are copied to TARGET. Will I do something wrong?

**Comment by outdoorbits** [comment]
It worked here... It's just what you did. Connect target like always, followed by the sources. Could you give more information? Filesystem types...

**Comment by outdoorbits** [comment]
Could you send the log file?

**Comment by jerzyks** [comment]
copy is in progress :-) this moment works - it looks like file system problem in one SD 64GB card - there was only few photos but property card says that is full. Formated and try copy again.

**Comment by jerzyks** [comment]
Ok, I did a few copies, formats, copies, formats and so on and everything looks very good. One time I had a situation when I had one photo on the card and LBB did not make a copy several times, but after taking additional photos it made a copy (vFAT). I'll try to make a few more copies to be sure. If necessary, I can send some logs from the equipment

If everything works as it should, you have landed in the perfect situation for me because in 4 days I will be going on holiday with my wife for 2 weeks where I will travel and, I hope, use LBB intensively. Without the multicopy option, I wouldn't even take LBB with me! THANKS A LOT!

**Comment by outdoorbits** [comment]
In my tests most times I had one file only on the sources. This has to work. Please report errors. Did you configure mails? If file have been missing, this should be documented in the mail.

**Comment by jerzyks** [comment]
email in github - yes
email in lbb raspberry - no = i have problems with initialize wifi and bluetooth, I don't know why yet...

**Comment by outdoorbits** [comment]
What about bluetooth? What's not working about mail?

**Comment by jerzyks** [comment]
> ave problems with initialize wifi and bluetooth, I don't know why yet...

ok - I'm stupid - I tried to configure the WiFi network when the device creates its own hotspot :-)))

**Comment by outdoorbits** [comment]
If you want to use your mobile add hotspot, I think you'll need a second device to configure. Comitup will remember this settings until you connect to another WiFi.

**Comment by jerzyks** [comment]
but then how to connect wifi to your own mobile hotspot?

I understand that the idea is to create a safe VPN between ETH and wifi, but it would probably be more popular to create a bridge between public wifi and your own hotspot created by LBB, and a long USB wifi card would be necessary for this

**Comment by outdoorbits** [comment]
Different problems. What do you want to do?

You have a vpn server at home, your smartphone in hotspot mode and lbb 
as client of the phone's hotspot and your vpn server?

**Comment by jerzyks** [comment]
I'll try to gather my thoughts:
- 1 situation: public non secured wifi connected to the first NIC LBB and on the second NIC connected smartphone (VPN, FIREWALL etc)
-2 situation: own WiFi hotspot with local operator's SIM - LBB connected to the first NIC, smartphone connected to the same WIFI, or smartphone connected to the second LBB NIC

Just to be clear - I'm not using any VPN for now, maybe in the future???

**Comment by outdoorbits** [comment]
I just can write my opinion, the real knowledge you will find here: 
http://davesteele.github.io/comitup/

1.: You use the public wifi as data connection to send mails and upload 
files. This connection is established by comitup. Lbb could establish a 
VPN while backing up to a cloud service. The second wifi on Rpi still 
stays in hotspot mode to control lbb via web UI.

2.: Your wifi hotspot is a independent device, yes? If you connect both 
to this wifi, lbb will get a new ip by your hotspot and if your hotspot 
allows connections between clients, you can access lbb by smartphone by 
this new IP (if your smartphone is inthis wifi also). Lbb's second wifi 
will still be in hotspot mode and you can connect your smartphone to 
this also. You even should have internet access for your phone through 
lbb and your independent hotspot. This way you also can pass portal 
pages of a wifi.

**Comment by jerzyks** [comment]
Then I don't know how comitup works, give me some time - I need to learn a bit

**Reply by outdoorbits** [reply]
Let's open a new discussion about Comitup is necessary. I'll close this one.
